apps/wallet-mobile/src/features/Swap/common/SwapFormProvider.tsx

```diff
@@ -28,59 +61,256 @@ export const SwapFormProvider = ({
+        (state.sellAmount.isTouched && state.buyAmount.isTouched && pool === undefined)
+      ? strings.notEnoughSupply
+      : undefined
+
+  const sellError =
+    noPoolError !== undefined
+      ? noPoolError
+      : !Quantities.isZero(sellQuantity) && state.buyAmount.isTouched && !hasSellBalance
+      ? strings.notEnoughBalance
+      : !Quantities.isZero(sellQuantity) && state.buyAmount.isTouched && !hasFeesBalance
+      ? strings.notEnoughFeeBalance
+      : undefined
+
+  const actions = React.useRef<SwapFormActions>({
     sellTouched: () => dispatch({type: 'sellTouched'}),
     buyTouched: () => dispatch({type: 'buyTouched'}),
     switchTouched: () => dispatch({type: 'switchTouched'}),
     poolTouched: () => dispatch({type: 'poolTouched'}),
     poolDefaulted: () => dispatch({type: 'poolDefaulted'}),
     resetTouches: () => dispatch({type: 'resetTouches'}),
+    setBuyInputValue: (value: string) => dispatch({type: 'setBuyInputValue', value}),
+    setSellInputValue: (value: string) => dispatch({type: 'setSellInputValue', value}),
+    setBuyAmountError: (error: string | undefined) => dispatch({type: 'setBuyAmountError', error}),
+    setSellAmountError: (error: string | undefined) => dispatch({type: 'setSellAmountError', error}),
   }).current
 
-  const context = React.useMemo(() => ({...state, ...actions}), [state, actions])
+  React.useEffect(() => {
+    if (state.buyAmount.isTouched && !buyInputRef?.current?.isFocused()) {
+      actions.setBuyInputValue(Quantities.format(orderData.amounts.buy.quantity, buyTokenInfo.decimals ?? 0))
+    }
+  }, [state.buyAmount.isTouched, orderData.amounts.buy.quantity, buyTokenInfo.decimals, actions])
 
-  return <TouchedContext.Provider value={context}>{children}</TouchedContext.Provider>
+  React.useEffect(() => {
+    if (state.sellAmount.isTouched && !sellInputRef?.current?.isFocused()) {
+      actions.setSellInputValue(Quantities.format(orderData.amounts.sell.quantity, sellTokenInfo.decimals ?? 0))
+    }
+  }, [actions, state.sellAmount.isTouched, orderData.amounts.sell.quantity, sellTokenInfo.decimals])
+
+  React.useEffect(() => {
+    if (sellError !== state.sellAmount.error) actions.setSellAmountError(sellError)
+  }, [actions, sellError, state.sellAmount.error])
+
+  React.useEffect(() => {
+    if (buyError !== state.buyAmount.error) actions.setBuyAmountError(buyError)
+  }, [actions, buyError, state.buyAmount.error])
+
+  const onChangeSellQuantity = React.useCallback(
+    (text: string) => {
+      try {
+        const [input, quantity] = Quantities.parseFromText(text, sellTokenInfo.decimals ?? 0, numberLocale)
+        sellQuantityChanged(quantity)
+        actions.setSellInputValue(text === '' ? text : input)
+      } catch (error) {
+        Alert.alert(strings.generalErrorTitle, strings.generalErrorMessage(error))
+      }
+    },
+    [actions, numberLocale, sellQuantityChanged, sellTokenInfo.decimals, strings],
+  )
+
+  const onChangeBuyQuantity = React.useCallback(
```

Comment:
please lift up the switch tokens into here, the logic goes form controlled = form action -> swap action.